### PR TITLE
Initial support for IPv6

### PR DESCRIPTION
**Pull request by trivialfis**
- Merge rabit socket into XGBoost.
- Dask interface support.
- Add test for socket.
- Add more compatibility code with mysys32, including correct linking, and enable most of
  the system function calls.
- Workaround windows includes in tests. dmlc filesystem includes `windows.h`, which
  conflicts with winsock2.h.

Current status:
This PR enables initial support for IPv6 with dask on Linux. Some effort is put into Windows support but we don't support distributed training on Windows in general. As for macos, RABIT allreduce returns error but I don't have a macos device for development or testing.

I ran some basic tasks using dask on GCP debian-based instance with dual stack and verified that IPv6 is indeed used.

https://github.com/dmlc/xgboost/issues/7725 .

### CODE REVIEW

**Comment by trivialfis** [comment]
The use of `collective` is to be consistent with https://github.com/dmlc/xgboost/pull/8057

File: rabit/src/allreduce_base.h [hunk 1]
```diff
@@ -201,8 +201,8 @@ class AllreduceBase : public IEngine {
     }
   };
   /*! \brief translate errno to return type */
-  inline static ReturnType Errno2Return() {
-    int errsv = utils::Socket::GetLastError();
+  static ReturnType Errno2Return() {
+    int errsv = errno;
```
Comment (trivialfis) [inline rabit/src/allreduce_base.h:205]: fix this.

**Comment by trivialfis** [comment]
@rongou Would you like to take a look into this?

File: doc/tutorials/dask.rst [hunk 2]
```diff
@@ -504,6 +503,35 @@ dask config is used:
...
 
+
+************
+IPv6 Support
+************
+
+.. versionadded:: 2.0.0
+
+XGBoost has initial IPv6 support for the dask interface on Linux. Due to most of the
+cluster support for IPv6 is partial (duo stack instead of IPv6 only), we require
```
Comment (rongou) [inline doc/tutorials/dask.rst:514]: dual stack

**Review by rongou — APPROVED** [review]

**Review by RAMitchell — COMMENTED** [review]
I can't really say much about the implementation as I don't have experience here.

It does seem a bit unusual to me that we are doing this ourselves and not using a library. Would using another framework for communication give us IPv6?

If you are confident with this PR you should go ahead with it.

**Comment by trivialfis** [comment]
@RAMitchell We use to treat rabit as a separate library. ;-) But as suggested by @rongou , we can switch to something well maintained for allreduce implementation.

**Pull request merged**